### ISSUE TITLE
Websocket advanced example

### ISSUE THREAD

**Issue by orangesoup** [gin-gonic/gin#461]
Hey!

My problem is a bit complex, but I hope someone can help me. I have an API server made with gin and an Angular app on client side. The authentication is done with JWT. So far so good, everything works.

Since websocket doesn't support `authorization` header I thougt I'll make the websocket urls like this: `/ws/:token/whatever` and use a middleware to validate the token. This is working, no problem there.

Here comes the problem though: Let's say I want to invalidate a user's tokens. I destroy them on server side, that's okay, but I also want to tell the user on the client side via websocket to remove the token from localStorage and then log him out. Obviously on a refresh, or state change, or whatever else the user would get `401` anyway, but I would like to sign him out still.

There is a basic websocket example in #51, but it doesn't solve my problem.
Basically: how can I write to an exact websocket from another API call?

Example: admin calls `/api/admin/user/sessions/clear`, posts the user id. This calls a function `ClearSessions()`. How can this `ClearSessions()` function also write to the user's websocket (`/ws/:token/whatever`)?

Thanks!

**Comment by nazwa** [comment]
You need to keep track of your sockets. Have a "global" synchronized map of account ID or token ID -> socket pointer. Every time new socket is created, just add it to the list.

This way you can control your sockets from the outside and if you close them through another api call your code can handle it nicely on both ends.

Also, don't send tokens in URL. Those can be intercepted super easily making your api insecure. You can send the authorization header in your call to open the socket as per usual and check it before socket is established.

```go
r.GET("/ws", func(c *gin.Context) {
   // CHECK AUTH HEADER HERE
(...)
// Open the socket...  this or whatever other way you are using.
    handler := websocket.Handler(EchoServer)
    handler.ServeHTTP(c.Writer, c.Req)
})
```

**Comment by orangesoup** [comment]
For the first part: could you pretty please write a really basic example on this "global" synchronized map would work?

And on the second part:
````go
r.GET("/ws", func(c *gin.Context) {
// CHECK AUTH HEADER HERE
}
````
How is this supposed to work? Even if I leave out the token from the url, I would end up with something like this on client side: 
````javascript
var dataStream = $websocket('/ws/whatever');
````

And you simply can't set headers on the websocket protocol (afaik), so I don't get how would you check the auth header there. :/

**Comment by nazwa** [comment]
Ha, sorry. You are absolutely right - forgot that sockets only send cookies around.

Keep an authentication  flag for every socket and require the first message from client to contain the authentication token. If client sends anything else than a valid authentication token and connection is still marked (using that flag) as unauthenticated then you can assume it's not a legitimate connection and close it or ban it.




As for the map, something like this:

```go 
type OpenSocketsStruct struct {
	sync.Mutex
	list map[in64]*websocket
}

func (p *OpenSocketsStruct) Add(id int64, ws *websocket) {
    p.Lock()
    defer p.Unlock()

    p.list[id] = ws
}

```

And then keep it somewhere that's accessible to your handlers.Obviously any reads and deletes also need to obtain the mutex as maps are not thread safe.

**Comment by nazwa** [comment]
You could even combine the two together and have the auth flag inside that global connection map.


```go
type ConnectionStruct struct {
    authenticated bool
    ws *websocket
}

type OpenSocketsStruct struct {
    sync.Mutex
    list map[in64]*ConnectionStruct 
}

func (p *OpenSocketsStruct) Add(id int64, ws *websocket) {
    p.Lock()
    defer p.Unlock()

    p.list[id] = &ConnectionStruct {
        authenticated: false,
        ws: websocket,
    }
}

```

**Comment by orangesoup** [comment]
Thanks! Got it working!
